README.md
# quick-highlight [![Build Status](https://travis-ci.org/t9md/atom-quick-highlight.svg?branch=master)](https://travis-ci.org/t9md/atom-quick-highlight)

- Highlight selected and multiple persisting highlight across visible editor.
![gif](https://raw.githubusercontent.com/t9md/t9md/f51b8e211e9ed8ed455053be52d5505da876b298/img/atom-quick-highlight.gif)

- Show found count on StatusBar.
![gif](https://raw.githubusercontent.com/t9md/t9md/a00e64b9dd85b851ad23c28e830f4a7d7dbe6dcf/img/atom-quick-highlight.png)

# Commands

- `quick-highlight:toggle` toggle highlight for selected or cursor word.
- `quick-highlight:clear` clear all highlight.

And following two operator for [vim-mode-plus](https://atom.io/packages/vim-mode-plus) user.
- `vim-mode-plus-user:quick-highlight`: Operator to highlight by text-object or motion.
- `vim-mode-plus-user:quick-highlight-word` Highlight cursor word, similar to `quick-highlight:toggle`, but well fit for vim's block cursor orientation. And `.` repeatable.

# Keymap

No default keymap.

e.g.
* general
```coffeescript
'atom-workspace atom-text-editor:not([mini])':
  'cmd-k m': 'quick-highlight:toggle'
  'cmd-k M': 'quick-highlight:clear'
```

* vim-mode-plus user
```coffeescript
'atom-text-editor.vim-mode-plus.normal-mode, atom-text-editor.vim-mode-plus.visual-mode':
  'space m': 'vim-mode-plus-user:quick-highlight-word'
  'space M': 'quick-highlight:clear'
  'g m': 'vim-mode-plus-user:quick-highlight'
```

## Modify highlight style

You can override style in you `style.less`.
See example below.

```less
@import "syntax-variables";

// For selection color
//=======================
atom-text-editor .quick-highlight.box-selection .region {
  border-width: 1px;
  background-color: transparent;
  border-color: @syntax-text-color;
}

// Make underline manual highlight prioritized(come front) over other highlight
//=======================
// Mixin to set z-index of quick-highlight manual color
.quick-highlight-z-index(@name, @value) {
  .quick-highlight.@{name} .region {
    z-index: @value;
  }
}

// quick-highlight use 0 to 7 color
//  for box style, use box-01 to box-07
//  for highlight style, use highlight-01 to highlight-07
.quick-highlight-z-index(underline-01, 1);
.quick-highlight-z-index(underline-02, 1);
.quick-highlight-z-index(underline-03, 1);
.quick-highlight-z-index(underline-04, 1);
.quick-highlight-z-index(underline-05, 1);
.quick-highlight-z-index(underline-06, 1);
.quick-highlight-z-index(underline-07, 1);
```

## vim-mode-plus operator

You can quick-highlight with combination of any motion, text-object.  
Since it's operator, yes can repeat by `.`.

e.g.
- `g m i w`: highlight `inner-word`.
- `g m i l`: highlight `inner-line`.
- `g m i'`: highlight `inner-single-quote`.
- `v 2 l g m`: highlight three visually selected character..

# Display found count on StatusBar

By default, when you highlight new text by `quick-highlight:toggle`, found count is displayed on StatusBar.  
You can configure CSS class to use with `countDisplayStyles`.  
See. `styleguide:show` for available style classes.  

e.g.
- Default: `badge icon icon-location`
- e.g. 2: `badge badge-error icon icon-bookmark`
- e.g. 3: `badge badge-success icon icon-light-bulb`
- e.g. 4: `btn btn-primary selected inline-block-tight`

# TODO

* [x] Improve default to dynamically change using [color-channel](http://lesscss.org/functions/#color-channel)?
* [ ] Highlight with RegExp.


CHANGELOG.md
## 0.13.0:
- Improve: `quick-highlight:toggle` now pick cursor word correctly when executed in editor in Dock.
- Maintenance: Standard linting, format.
- Style: Change default value for `countDisplayStyles` which is used for statusbar icon style.
  - New: `icon icon-location`
  - Old: `badge icon icon-location`(If you like older style, set this manually)

## 0.12.0:
- Maintenance: Catchup vmp changes, no longer use service.Base, use getClass instead.

## 0.11.1:
- Update `registerCommandFromSpec` call to conform to vmp changes.
  - Detail: https://github.com/t9md/atom-vim-mode-plus/pull/958

## 0.11.0:
- Maintenance:
  - Convert Coffee-to-JS
  - Also make vmp commands compatible with upcoming ES6-class-based-vmp-operations.

## 0.10.0:
- Improve: Improve package activation perf by reducing disk-IO on startup.
  - Activation time diff: from 20ms to 5ms in my env.

## 0.9.0:
- Reduce activation time for `vim-mode-plus` was activated by new `registerCommandFromSpec` service.

## 0.8.1:
- Fix: No longer set `z-index` of all quick-highlight decoration to avoid covering underlying text.
  - If you want to tweak `z-index`, see example put in `README.md`
- Improve: Use better transparent color calculation for selection highlight background color. #16

## 0.8.0:
- Fix: Avoid flickering selection highlightSelection in `vim-mode-plus`.
- New: Add service to integrate `minimap`

## 0.7.2:
- Improve: Early skip calling highlightSelection to exclude cursor movement from debounce calculation.

## 0.7.0:
- Completely rewritten.
- No longer refresh every `onDidChangeActivePaneItem`.
- No longer do tuning to render "only-visible-screen-area". Instead, it create marker/decoration for all matching keyword in editor.
- Default highlight style is now `underline`

## 0.6.2
- Improve: Update selection-highlight style(`box-selection`).
- Doc: Add TIPS to modify highlight style in README.md.

## 0.6.1
- Set `z-index` for box highlight to win selection marker.

## 0.6.0
- Fix to follow upstream(`vim-mode-plus`) change.

## 0.5.0
- Fix: Deprecation warning from Atom1.9.0-beta0.
- Improve: Refactoring

## 0.4.0
- Add operator command for vim-mode-plus, `vim-mode-plus-user:quick-highlight`.
- Improve accuracy for picking word under cursor by using `selection.selectWord()`.

## 0.3.10 - FIX
- Guard for `editorElement.getVisibleRowRange()` return `null` at initialization #7

## 0.3.9 - Improve
- New: config parameter highlightSelectionDelay suggested by @PaulPorfiroff.

## 0.3.8 - FIX
- No longer use TextEditor constructor directly work around warning from Atom 1.2.0

## 0.3.7 - New feature
- New highlightSelectionExcludeScopes setting allow you to exclude highlight selection by scopes.

## 0.3.6 - FIX
- Explicitly pass editor reference to highlightSelection() so that rapid pane change at Atom startup don't cause error.

## 0.3.5 - New feature
- New: highlight selection enabled by default, you can disable from settings.

## 0.3.4 - FIX
- Remove unnecessary event observation for onDidChangeScrollLeft

## 0.3.3 - FIX
- [FIX] Uncaught TypeError: Invalid Point: (NaN, 0) #2, from Atom v1.1.0
- [FIX] Deprecation waring from Atom v1.1.0.

## 0.3.2 - Add spec
- More spec coverage: status-bar indicator.
- Improve default color. now change color based on @syntax-background-color

## 0.3.1 - Add spec
- Add spec
- Precise check for decorationsByEditor has keywords.

## 0.3.0 - Improve
- Completely rewrite.
- Now decoration is created only for visible area. Greatly improve performance.

## 0.2.2 - Improve

## 0.1.12 - Doc
- Update readme to follow vim-mode's rename from command-mode to normal-mode

## 0.1.11 - BUG FIX
* [FIX] Could not highlight for Object's property like `hasOwnProperty`, `__defineGetter__`.

## 0.1.10 - Improve
* [FIX] box color broken in TILE border.

## 0.1.9 - Improve
* Doc update.

## 0.1.8 - StatusBar update.
* Now display found count on StatusBar.

## 0.1.7 - Refactoring
* Clear editor subscriptions.
* Refactoring
* Remove unused keymap directory

## 0.1.6 - Fix deprecated API, atom/atom#6867
* Use TextEditor::getLastSelection() instead of getSelection()

## 0.1.5 - Improve
* Clean up

## 0.1.3 - Improve
* Now `box` is default for config `quick-highlight.decorate`.
* Refactoring
* Refresh on Pane become active(=onDidChangeActivePaneItem()).
* Restore old cursorPosition.
* Now highlight refreshed on TextEditor::onDidChange

## 0.1.2 - Box decoration support
* Now user can choose 'highlight', 'box' for decoration preference.

## 0.1.1 - Fix
* FIX: Didn't work correctly.
* Cleanup

## 0.1.0 - First Release
* Highlight within current editor.


lib/main.js
const {CompositeDisposable, Emitter} = require('atom')

let KeywordManager

module.exports = {
  activate (state) {
    this.emitter = new Emitter()
    this.toggle = this.toggle.bind(this)
    const toggle = this.toggle

    this.subscriptions = new CompositeDisposable(
      atom.commands.add('atom-text-editor:not([mini])', {
        'quick-highlight:toggle' () {
          toggle(getCursorWord(this.getModel()))
        },
        'quick-highlight:clear': () => this.keywordManager && this.keywordManager.clear()
      }),
      atom.config.observe('quick-highlight.highlightSelection', value => {
        if (value) this.getKeywordManager() // To initialize
      })
    )
  },

  deactivate () {
    if (this.keywordManager) this.keywordManager.destroy()
    this.subscriptions.dispose()
    this.keywordManager = null
    this.subscriptions = null
  },

  getKeywordManager () {
    if (!KeywordManager) KeywordManager = require('./keyword-manager')
    if (!this.keywordManager) {
      this.keywordManager = new KeywordManager(this.emitter)
      this.setStatusBarService()
    }
    return this.keywordManager
  },

  toggle (keyword) {
    this.getKeywordManager().toggle(keyword)
  },

  onDidChangeHighlight (fn) {
    return this.emitter.on('did-change-highlight', fn)
  },

  provideQuickHighlight () {
    return {onDidChangeHighlight: this.onDidChangeHighlight.bind(this)}
  },

  setStatusBarService () {
    if (this.statusBarService && this.keywordManager) {
      this.keywordManager.setStatusBarService(this.statusBarService)
    }
  },

  consumeStatusBar (service) {
    this.statusBarService = service
    this.setStatusBarService()
  },

  consumeVim ({getClass, registerCommandsFromSpec}) {
    this.subscriptions.add(
      registerCommandsFromSpec(['QuickHighlight', 'QuickHighlightWord'], {
        prefix: 'vim-mode-plus-user',
        loader: () => require('./load-vmp-commands')(getClass, this.toggle)
      })
    )
  }
}

function getCursorWord (editor) {
  const selection = editor.getLastSelection()
  const selectedText = selection.getText()
  if (selectedText) {
    return selectedText
  } else {
    const cursorPosition = selection.cursor.getBufferPosition()
    selection.selectWord()
    const word = selection.getText()
    selection.cursor.setBufferPosition(cursorPosition)
    return word
  }
}


lib/status-bar-manager.js
module.exports = class StatusBarManager {
  constructor () {
    this.span = document.createElement('span')

    this.element = document.createElement('div')
    this.element.id = 'status-bar-quick-highlight'
    this.element.className = 'block'
    this.element.appendChild(this.span)

    this.container = document.createElement('div')
    this.container.className = 'inline-block'
    this.container.appendChild(this.element)
  }

  initialize (statusBarService) {
    this.statusBarService = statusBarService
  }

  update (count) {
    this.span.className = atom.config.get('quick-highlight.countDisplayStyles')
    this.span.textContent = count
    this.element.style.display = 'inline-block'
  }

  clear () {
    this.element.style.display = 'none'
  }

  attach () {
    const displayPosition = atom.config.get('quick-highlight.countDisplayPosition')

    this.tile = this.statusBarService[`add${displayPosition}Tile`]({
      item: this.container,
      priority: atom.config.get('quick-highlight.countDisplayPriority')
    })
  }

  detach () {
    if (this.tile) this.tile.destroy()
  }
}


lib/quick-highlight-view.js
const _ = require('underscore-plus')
const {CompositeDisposable} = require('atom')

function matchScopes ({classList}, scopes = []) {
  return scopes.some(scope => scope.split('.').every(name => classList.contains(name)))
}

function getConfig (name) {
  return atom.config.get(`quick-highlight.${name}`)
}

function getVisibleEditors () {
  return atom.workspace
    .getPanes()
    .map(pane => pane.getActiveEditor())
    .filter(v => v)
}

function isVisibleEditor (editor) {
  return getVisibleEditors().includes(editor)
}

function isActiveEditor (editor) {
  return atom.workspace.getActiveTextEditor() === editor
}

// - Refresh onDidChangeActivePaneItem
// - But dont't refresh invisible editor
// - Update statusbar count on activeEditor was changed
// - Clear marker for invisible editor?: Decide to skip to avoid clere/re-render.
// - Update only keyword added/remove: Achived by diffing.
class QuickHighlightView {
  static initClass () { this.viewByEditor = new Map() } // prettier-ignore
  static register (view) { this.viewByEditor.set(view.editor, view) } // prettier-ignore
  static unregister (view) { this.viewByEditor.delete(view.editor) } // prettier-ignore
  static destroyAll () { this.viewByEditor.forEach(view => view.destroy()) } // prettier-ignore
  static clearAll () { this.viewByEditor.forEach(view => view.clear()) } // prettier-ignore

  static refreshVisibles () {
    const editors = getVisibleEditors()
    for (const editor of editors) {
      const view = this.viewByEditor.get(editor)
      if (view) {
        view.refresh()
      }
    }
  }

  constructor (editor, {keywordManager, statusBarManager, emitter}) {
    this.editor = editor
    this.keywordManager = keywordManager
    this.statusBarManager = statusBarManager
    this.emitter = emitter
    this.markerLayerByKeyword = new Map()

    this.highlightSelection = this.highlightSelection.bind(this)
    let highlightSelection
    this.disposables = new CompositeDisposable(
      atom.config.observe('quick-highlight.highlightSelectionDelay', delay => {
        highlightSelection = _.debounce(this.highlightSelection, delay)
      }),
      this.editor.onDidDestroy(() => this.destroy()),
      this.editor.onDidChangeSelectionRange(({selection}) => {
        if (selection.isEmpty()) {
          this.clearSelectionHighlight()
        } else {
          highlightSelection(selection)
        }
      }),
      this.editor.onDidStopChanging(() => {
        this.clear()
        if (isVisibleEditor(this.editor)) {
          this.refresh()
        }
      })
    )

    QuickHighlightView.register(this)
  }

  needSelectionHighlight (text) {
    return (
      getConfig('highlightSelection') &&
      !matchScopes(this.editor.element, getConfig('highlightSelectionExcludeScopes')) &&
      text.length >= getConfig('highlightSelectionMinimumLength') &&
      !/\n/.test(text) &&
      /\S/.test(text)
    )
  }

  highlightSelection (selection) {
    this.clearSelectionHighlight()
    const keyword = selection.getText()
    if (this.needSelectionHighlight(keyword)) {
      this.markerLayerForSelectionHighlight = this.highlight(keyword, 'box-selection')
    }
  }

  clearSelectionHighlight () {
    if (this.markerLayerForSelectionHighlight) this.markerLayerForSelectionHighlight.destroy()
    this.markerLayerForSelectionHighlight = null
  }

  highlight (keyword, color) {
    const markerLayer = this.editor.addMarkerLayer()
    this.editor.decorateMarkerLayer(markerLayer, {type: 'highlight', class: `quick-highlight ${color}`})
    const regex = new RegExp(`${_.escapeRegExp(keyword)}`, 'g')
    this.editor.scan(regex, ({range}) => markerLayer.markBufferRange(range, {invalidate: 'inside'}))

    if (markerLayer.getMarkerCount()) {
      this.emitter.emit('did-change-highlight', {
        editor: this.editor,
        markers: markerLayer.getMarkers(),
        color: color
      })
    }
    return markerLayer
  }

  getDiff () {
    const masterKeywords = this.keywordManager.getKeywords()
    const currentKeywords = [...this.markerLayerByKeyword.keys()]
    const newKeywords = _.without(masterKeywords, ...currentKeywords)
    const oldKeywords = _.without(currentKeywords, ...masterKeywords)

    if (newKeywords.length || oldKeywords.length) {
      return {newKeywords, oldKeywords}
    }
  }

  render ({newKeywords, oldKeywords}) {
    // Delete
    for (const keyword of oldKeywords) {
      this.markerLayerByKeyword.get(keyword).destroy()
      this.markerLayerByKeyword.delete(keyword)
    }

    // Add
    for (const keyword of newKeywords) {
      const color = this.keywordManager.getColorForKeyword(keyword)
      if (color) {
        const decorationColor = atom.config.get('quick-highlight.decorate')
        const markerLayer = this.highlight(keyword, `${decorationColor}-${color}`)
        this.markerLayerByKeyword.set(keyword, markerLayer)
      }
    }
  }

  clear () {
    this.markerLayerByKeyword.forEach(layer => layer.destroy())
    this.markerLayerByKeyword.clear()
  }

  refresh () {
    const diff = this.getDiff()
    if (diff) {
      this.render(diff)
    }
    this.updateStatusBarIfNecesssary()
  }

  updateStatusBarIfNecesssary () {
    if (getConfig('displayCountOnStatusBar') && isActiveEditor(this.editor)) {
      this.statusBarManager.clear()
      const keyword = this.keywordManager.latestKeyword

      const layer = this.markerLayerByKeyword.get(keyword)
      const count = layer ? layer.getMarkerCount() : 0
      if (count) this.statusBarManager.update(count)
    }
  }

  destroy () {
    this.disposables.dispose()
    this.clear()
    this.clearSelectionHighlight()
    QuickHighlightView.unregister(this)
  }
}

QuickHighlightView.initClass()
module.exports = QuickHighlightView


lib/keyword-manager.js
const {Emitter, CompositeDisposable} = require('atom')
const QuickHighlightView = require('./quick-highlight-view')
const StatusBarManager = require('./status-bar-manager')

const Colors = {
  colorNumbers: ['01', '02', '03', '04', '05', '06', '07'],
  next () {
    this.index = (this.index + 1) % this.colorNumbers.length
    return this.colorNumbers[this.index]
  },
  reset () {
    this.index = -1
  }
}

module.exports = class KeywordManager {
  constructor (mainEmitter) {
    this.colorByKeyword = new Map()
    this.reset()

    this.latestKeyword = null

    this.emitter = new Emitter()
    this.statusBarManager = new StatusBarManager()

    this.disposables = new CompositeDisposable(
      atom.workspace.observeTextEditors(editor => {
        /* eslint-disable no-new */
        new QuickHighlightView(editor, {
          keywordManager: this,
          statusBarManager: this.statusBarManager,
          emitter: mainEmitter
        })
        /* eslint-enable */
      }),
      atom.workspace.onDidChangeActivePaneItem(() => QuickHighlightView.refreshVisibles()),
      atom.config.observe('quick-highlight.decorate', newValue => {
        QuickHighlightView.clearAll()
        QuickHighlightView.refreshVisibles()
      })
    )
  }

  reset () {
    this.colorByKeyword.clear()
    Colors.reset()
  }

  setStatusBarService (service) {
    this.statusBarManager.initialize(service)
    this.statusBarManager.attach()
  }

  toggle (keyword) {
    if (this.colorByKeyword.has(keyword)) {
      this.colorByKeyword.delete(keyword)
    } else {
      this.colorByKeyword.set(keyword, Colors.next())
      this.latestKeyword = keyword
    }

    QuickHighlightView.refreshVisibles()
  }

  getColorForKeyword (keyword) {
    return this.colorByKeyword.get(keyword)
  }

  getKeywords () {
    return [...this.colorByKeyword.keys()]
  }

  clear () {
    this.reset()
    QuickHighlightView.clearAll()
  }

  destroy () {
    QuickHighlightView.destroyAll()

    this.disposables.dispose()
    this.disposables = null

    this.statusBarManager.detach()
    this.statusBarManager = null
  }
}
